nixos/{networkd,dhcpcd}: remove udev-settle hack

rnhmjoj opened the pull request:
###### Motivation for this change

systemd-udev-settle is a terrible hack[1] and should never[2] ever[3]
used, seriously it's very bad. It was used as a stop-gap solution for
issue #39069, but thanks to PR #79532 it can be removed now.

[1]: https://github.com/systemd/systemd/issues/7293#issuecomment-592941764
[2]: https://github.com/NixOS/nixpkgs/issues/73095
[3]: https://github.com/NixOS/nixpkgs/issues/107341


###### Things done

- Tested via:
  - [x]  `nixosTests.networking.networkd`
  - [x]  `nixosTests.networking.scripted`
  - [x] `nixosTests.predictable-interface-names` 
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

rnhmjoj committed:
nixos/{networkd,dhcpcd}: remove udev-settle hack

systemd-udev-settle is a terrible hack[1] and should never[2] ever[3]
used, seriously it's very bad. It was used as a stop-gap solution for
issue #39069, but thanks to PR #79532 it can be removed now.

[1]: https://github.com/systemd/systemd/issues/7293#issuecomment-592941764
[2]: https://github.com/NixOS/nixpkgs/issues/73095
[3]: https://github.com/NixOS/nixpkgs/issues/107341

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/boot-faster-by-disabling-udev-settle-and-nm-wait-online/6339/4

flokli commented:
Shower thought: Do we copy and apply all udev rules into initrd?

I'm a bit worried about `70-persistent-net.rules`-style udev rules not getting applied before dhcpcd starts doing its thing. Such a rule is mentioned as an example in `nixos/modules/services/hardware/udev.nix` for `services.udev.extraRules`

The `systemd-udev-settle.service` might be what's preventing this from happening for now.

We should add a test for this, to ensure it keeps working.

peterhoeg commented:
There is some additional information here: https://github.com/NixOS/nixpkgs/pull/25311

rnhmjoj commented:
I run more tests today and everything looks fine:
```
nixosTests.containers-restart_networking
nixosTests.initrdNetwork
nixosTests.installer.bcache
nixosTests.installer.btrfsSimple
nixosTests.installer.btrfsSubvolDefault
nixosTests.installer.btrfsSubvols
nixosTests.installer.encryptedFSWithKeyfile
nixosTests.installer.grub1
nixosTests.installer.luksroot
nixosTests.installer.luksroot-format1
nixosTests.installer.luksroot-format2
nixosTests.installer.lvm
nixosTests.installer.separateBoot
nixosTests.installer.separateBootFat
nixosTests.installer.simple
nixosTests.installer.simpleLabels
nixosTests.installer.simpleProvided
nixosTests.installer.simpleSpecialised
nixosTests.installer.simpleUefiGrub
nixosTests.installer.simpleUefiGrubSpecialisation
nixosTests.installer.simpleUefiSystemdBoot (fails as of  7b72e86481c09b25b44fc319aff596c828b5f95f, pass after rebase)
nixosTests.installer.swraid
nixosTests.installer.zfsroot
nixosTests.networking.networkd.bond
nixosTests.networking.networkd.bridge
nixosTests.networking.networkd.dhcpOneIf
nixosTests.networking.networkd.dhcpSimple
nixosTests.networking.networkd.link
nixosTests.networking.networkd.loopback
nixosTests.networking.networkd.privacy
nixosTests.networking.networkd.routes
nixosTests.networking.networkd.sit
nixosTests.networking.networkd.static
nixosTests.networking.networkd.virtual
nixosTests.networking.networkd.vlan
nixosTests.networking.scripted.bond
nixosTests.networking.scripted.bridge
nixosTests.networking.scripted.dhcpOneIf
nixosTests.networking.scripted.dhcpSimple
nixosTests.networking.scripted.link
nixosTests.networking.scripted.loopback
nixosTests.networking.scripted.macvlan
nixosTests.networking.scripted.privacy
nixosTests.networking.scripted.routes
nixosTests.networking.scripted.sit
nixosTests.networking.scripted.static
nixosTests.networking.scripted.virtual
nixosTests.networking.scripted.vlan
```

flokli commented:
@rnhmjoj can you add a test for the `services.udev.extraRules` rename example into `nixos/tests/networking.nix`? I assume you could just add another interface (via the `vlan` argument), and ensure it gets renamed via the udev rule.

rnhmjoj commented:
@flokli Thank you for the suggestion: there is indeed a problem here. I wrote this small test:
```nix
 {
   rename = {
      name = "RenameInterface";
      machine = { pkgs, ... }: {
        virtualisation.vlans = [ 1 ];
        networking = {
          useNetworkd = networkd;
          useDHCP = false;
        };
        services.udev.extraRules = ''
          SUBSYSTEM=="net", ACTION=="add", DRIVERS=="?*", ATTR{address}=="52:54:00:12:01:01", KERNEL=="eth*", NAME="custom_name"
        '';
      };
      testScript = ''
        machine.succeed("udevadm settle")
        print(machine.succeed("ip link show dev custom_name"))
      '';
    };
}
```
The "scripted" variant works, but the renaming fails when netword is enabled:
 
    systemd-udevd[481]: eth1: Failed to rename network interface 3 from 'eth1' to 'custom_name': Device or resource busy

I'm not sure what this means...

flokli commented:
The linked https://github.com/NixOS/nixpkgs/pull/79532 kinda does something similar this:

[We copy `${udev}/lib/systemd/network/*.link` into the initrd](https://github.com/NixOS/nixpkgs/pull/79532/files#diff-b636d2e49e098b09a6f1b276ad981f0772cc9f93b23c3d23b3bdff54cd8fb287R207), and also some udev rules related to interface renaming, so the renaming happens early, before interfaces are up (and we don't race with networkd or scripted networking also configuring interfaces).

I feel like we need to copy more of the `.link` files into initrd, and the udev extraRules as well. cc-ing @fpletz, @Ma27, @emilazy and @mweinelt from #79532 for opinions.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-in-distress/3604/39

rnhmjoj committed:
nixos/stage-1: install networkd link files

Renaming an interface must be done in stage-1: otherwise udev will
report the interface as ready and network daemons (networkd, dhcpcd,
etc.) will bring it up. Once up the interface can't be changed and the
renaming will fail.

Note: link files are read directly by udev, so they can be used even
without networkd enabled.

rnhmjoj committed:
nixos/tests/networking: test interface renaming

rnhmjoj committed:
nixos/udev: add option to install rules in initrd

Note: this moves the example rule used to rename network interfaces in
the new udev.initrdRules option, which is required since 115cdd1c.

flokli commented on nixos/doc/manual/configuration/renaming-interfaces.xml:
```diff
@@ -0,0 +1,67 @@
+  to assign names to network interfaces. This means that by default
+  cards are not given the traditional names like
+  <literal>eth0</literal> or <literal>eth1</literal>, whose order can
+  change unpredictably across reboots. Instead, using on physical
```
```suggestion
  change unpredictably across reboots. Instead, using physical
```

or "relying on"/"taking into consideration"…

flokli commented on nixos/doc/manual/configuration/renaming-interfaces.xml:
```diff
@@ -0,0 +1,67 @@
+  <programlisting>
+ <link linkend="opt-systemd.network.links">systemd.network.links."10-wan"</link> = {
+   matchConfig.MACAddress = "52:54:00:12:01:01";
+   linkConfig.Name = "wa";
```
```suggestion
   linkConfig.Name = "wan";
```

rnhmjoj replied:
Fixed.

flokli commented on nixos/doc/manual/configuration/renaming-interfaces.xml:
```diff
@@ -0,0 +1,67 @@
+ };
+  </programlisting>
+  <para>
+   Note that links are directly read by udev and will work even if
```
```suggestion
   Note that links are directly read by udev, not networkd, and will work even if
```

rnhmjoj replied:
Done.

flokli commented on nixos/doc/manual/configuration/renaming-interfaces.xml:
```diff
@@ -0,0 +1,67 @@
+  </programlisting>
+
+  <warning><para>
+    Unless the initrd has been entirely disabled, the rule must be installed
```
I wasn't aware you /CAN/ disable the initrd entirely, so this sounds a bit odd.

rnhmjoj replied:
It can technically be done since https://github.com/NixOS/nixpkgs/pull/80114, which simply makes configurable what was hardcoded to (!isContainer), but I'm not sure how safe it is on a real system. You probably have to know what you're doing pretty well.

flokli replied:
I'm not even sure if you can rename network interfaces from inside a nixos container. I'd just drop the "Unless …," part - in cases where you manually disable the initrd, one can expect people to take a closer look if initrd udev rules are applied or not ;-)

rnhmjoj replied:
Dropped.

flokli commented on nixos/doc/manual/release-notes/rl-2105.xml:
```diff
@@ -91,6 +91,16 @@
+     If you are using <option>services.udev.extraRules</option> to assign
+     custom names to network interfaces, this may stop working due to a change
+     in the initialisation of dhcpcd and systemd networkd. To avoid this, either
+     switch to <option>services.udev.initrdRules</option> or see the new
```
```suggestion
     move them to <option>services.udev.initrdRules</option> or see the new
```

rnhmjoj replied:
Done.

flokli commented on nixos/modules/services/hardware/udev.nix:
```diff
@@ -202,12 +202,22 @@ in
+        description = ''
+          <command>udev</command> rules to include in the initrd. They'll
+          be written into file <filename>99-local.rules</filename>. Thus
+          they are read and applied after all other rules.
```
Are they also added to the udev rules on the real system?

What does "all other rules" here refer to?

rnhmjoj replied:
There's a bunch of hardcoded rules in nixos/modules/system/boot/stage-1.nix. Mostly udev, lvm and some display manager thing.

flokli replied:
Okay, so the initrd udev rules are *only* added to the initrd, not the main system?

rnhmjoj replied:
Correct, I'll write it more clearly.

flokli commented on nixos/modules/services/hardware/udev.nix:
```diff
@@ -202,12 +202,22 @@ in
+
+      extraRules = mkOption {
+        default = "";
+        type = types.lines;
```
We probably should still keep an example here. Maybe

```
ENV{ID_VENDOR_ID}=="046d", ENV{ID_MODEL_ID}=="0825", ENV{PULSE_IGNORE}="1"
```

?

rnhmjoj replied:
Good idea, I'll grab this example.

rnhmjoj replied:
Added.

rnhmjoj committed:
nixos/docs: add section on renaming interfaces

rnhmjoj committed:
nixos/release-notes: warn on interface renaming

erictapen commented:
@rnhmjoj I'm using this for a while now ontop of `nixos-20.09`. What do you think about backporting this?

rnhmjoj commented:
@erictapen Yes, unfortunately it can't be easily backported.
If you're not using gdm or zfs, it should be mostly safe to do `systemd-udev-settle.enable = false` and disable udev-settle entirely.

arianvp commented on nixos/modules/services/hardware/udev.nix:
```diff
@@ -202,12 +202,26 @@ in
       };
 
-      extraRules = mkOption {
+      initrdRules = mkOption {
```
Hmmph not particularily fond of having this option defined here.  This module is for configuring systemd-udevd in **stage 2**; not **stage 1** 

 how about instead of having an `initrdRules` here; we move this option to `stage-1.nix` and name it `boot.initrd.extraUdevRules` ?  It's a bit weird to have the the option for initrd defined here as `udevd` in initrd and `udevd` in stage-2 might be two completely different `udevd` implementations at this moment.

I'm working on an alternative initrd implementation using systemd. But it's hard to introduce this change as now the `udev` module in stage-2 has a hard dependency on the `stage-1.nix` module.  Because now  the `stage-2` udev module calls out to `stage-1.nix`  I can not disable `stage-1.nix` in `disabledModules`

arianvp replied:
I'll send a draft PR with these suggested changes and i'll backlink it here.

flokli replied:
I think renaming this option is fine - as long as we do this before the 20.05 branchoff, it probably doesn't even need the `mkRenamedOption` alias.
